[WIP] 30 create input field

nicolechung opened the pull request:
Currently: a very early draft version.

Todos
- [ ] clean up git history with proper commit messages
- [ ] Add images of docs
- [ ] docs: add error example
- [ ] docs: add focus state example
- [ ] docs: add `number` type example
- [ ] docs: add `time` example
- [ ] docs: add `week` example
- [ ] docs: add `month` example
- [ ] tests: it renders with label and hint
- [ ] tests: it renders with proper error message
- [ ] tests: it can have `...attributes`

nicolechung commented on ember-toucan-core/src/components/form/input-field.hbs:
```diff
@@ -0,0 +1,26 @@
+<div class="flex flex-col {{if @isDisabled 'text-disabled cursor-not-allowed'}}">
```
Note: none of this is the finished styling currently, as we are waiting on user and design research. Styling will be for an upcoming iteration.

nicolechung commented on test-app/tests/integration/components/input-field-test.gts:
```diff
@@ -0,0 +1,90 @@
+
+    const label = 'label'
+    const input = 'input';
+    const error = 'div.text-critical';
```
**Question:** wondering if I should add `data-test` selectors in the https://github.com/CrowdStrike/ember-toucan-core/pull/35/files#diff-a8dfe5662c125f37bbd67d22d9112a123c0ffac52f7254a0fa217dd019452071 to get around these horribly specific class selectors?

ynotdraw commented on ember-toucan-core/src/components/form/input-field.hbs:
```diff
@@ -0,0 +1,32 @@
+
+    <field.Control class="flex">
+      <input
+        class="m-1 h-6 w-0 flex-grow bg-transparent p-1
```
Do we want the margin here via `m-1`? That'll push elements around a bit.

I tried getting _pretty_ close to what we are going for in with the textarea, which I think is what we'll mostly want here https://github.com/CrowdStrike/ember-toucan-core/pull/36/files#diff-0cf8ecb43c00a4f64c9e8710b853499367e0130793dffdb772c0bb4ef9a4dd0eR13


Something like this may be close:
```hbs
class="block h-6 p-1 bg-overlay-1 focus:outline-none rounded-sm shadow-focusable-outline focus:shadow-focus-outline {{if @isDisabled 'text-disabled' 'text-titles-and-attributes'}}"
```

I also left off other soon™ disabled styles, but happy to align with whatever so we remain consistent as to make it easier on us later when we go and update the styles to match post-testing!

ynotdraw commented on ember-toucan-core/src/components/form/input-field.ts:
```diff
@@ -0,0 +1,18 @@
+interface ToucanFormInputFieldComponentSignature {
+  Element: HTMLInputElement;
+  Args: {
+    label?: string;
```
Should we make `label` required for a11y reasons? (Even if we eventually allow it to be visually hidden)

```suggestion
    label: string;
```

nicolechung replied:
I totally missed this from the notes...need to add an assert here too (and test for it)

ynotdraw commented on ember-toucan-core/src/components/form/input-field.ts:
```diff
@@ -0,0 +1,18 @@
+  };
+}
+
+// eslint-disable-next-line ember/no-empty-glimmer-component-classes
+export default class ToucanFormInputFieldComponent extends Component<ToucanFormInputFieldComponentSignature> {}
```
If you're not using a constructor here, you could do this instead:

```js
import templateOnlyComponent from '@ember/component/template-only';

export default templateOnlyComponent<ToucanFormInputFieldComponentSignature>();
```

but if we do plan on adding a constructor eventually it's probably moot

nicolechung replied:
This should be a class now

ynotdraw commented on pnpm-lock.yaml:
```diff
@@ -8607,7 +8607,7 @@ packages:
       enhanced-resolve: 5.12.0
       eslint: 8.33.0
-      eslint-plugin-import: 2.27.5_fhrhggiruboehhi74bczdxbcti
+      eslint-plugin-import: 2.27.5_nowqz4jutkd4a233czbfk7jsgu
```
Do you know why this changed by chance? Since dependencies weren't adjusted in a `package.json` I wouldn't have expected any `ppm-lock.yaml` file changes

nicolechung replied:
Maybe it's because I did `pnpm i` without `--frozen-lockfile`??

https://pnpm.io/cli/install#--frozen-lockfile

nicolechung replied:
I thought that was only needed in CI though

nicolechung replied:
any ideas @ynotdraw ? I wonder (since I removed the dependency) if I should just revert this change. Let me try that...

ynotdraw commented on docs/components/input-field/index.md:
```diff
@@ -0,0 +1,133 @@
+
+## UI States
+
+### InputField with Label
```
I like the named headings here for UI states.  I'll update my PR to follow!

(For others: @nicolechung and I brainstormed how this may be a good idea if we want to add visual regression testing.  Then we can see the component in every expected state and visually diff things)

ynotdraw reviewed (commented):
All pretty minor comments -  This is exciting to see come together 🎉  🚀 .  Happy to pair tomorrow as well if we want to land on some consistency ideas/templates between our two PRs!

ynotdraw commented on ember-toucan-core/src/components/form/input-field.hbs:
```diff
@@ -0,0 +1,32 @@
+        aria-describedby="{{if @hint field.hintId}} {{if @error field.errorId}}"
+        aria-invalid={{if @error "true"}}
+        disabled={{@isDisabled}}
+        ...attributes
```
Kind of wondering if it's worth it for us to set `type="text"` by default?  Then folks can override since we `...attributes`.  What do you think?

simonihmig reviewed (commented):
I saw the request for review here only after the request for #36, so reviewed that first. But a couple of comments I left there apply to this in similar ways. So I am not duplicating them here. Instead maybe we could discuss them over there, and depending on the outcomes apply changes (if there are any) here accordingly?

ynotdraw commented on test-app/tests/integration/components/input-field-test.gts:
```diff
@@ -0,0 +1,138 @@
+  });
+
+  test('it accepts @value and @onChange', async function (assert) {
+    const onChangeCallback = sinon.spy();
```
Instead of adding `sinon` as a dependency you can use `assert.step` instead! https://github.com/CrowdStrike/ember-toucan-core/blob/main/test-app/tests/integration/components/button-test.gts#L131

nicolechung commented on ember-toucan-core/src/components/form/controls/input.ts:
```diff
@@ -0,0 +1,24 @@
+interface ToucanFormControlsInputComponentSignature {
+  Element: HTMLInputElement;
+  Args: {
+    isDisabled?: boolean;
```
If anyone has any advice as to why this is erroring: https://github.com/CrowdStrike/ember-toucan-core/actions/runs/4247201741/jobs/7384984044 would appreciate it!

nicolechung commented on docs/components/input-field/index.md:
```diff
@@ -0,0 +1,133 @@
+# Input field
```
I can adjust my "field" doc to match yours...let me take a look.

nicolechung replied:
Should be updated now in this commit: https://github.com/CrowdStrike/ember-toucan-core/pull/35#:~:text=match%20TextareaField%20docs-,6428f0f,-fix%3A%20Input%20Control

nicolechung commented:
> review here only after the request for #36, so reviewed that first. But a couple of comments I left there apply to this in similar ways. So I am not duplicating them here. Instead maybe we could discuss them over there, and depending on the outcomes apply changes (if there are any) here accordingly?

I ended up pairing with @ynotdraw so hopefully this addresses the issues mentioned.

ynotdraw commented on test-app/package.json:
```diff
@@ -112,6 +112,7 @@
     "prettier-plugin-ember-template-tag": "^0.3.0",
     "qunit": "^2.19.3",
     "qunit-dom": "^2.0.0",
+    "sinon": "^15.0.1",
```
We shouldn't need this anymore which should remove it from the lock file as well! 🎉

ynotdraw commented on docs/components/input/index.md:
```diff
@@ -0,0 +1,40 @@
+# Input 
+
+Provides a Toucan-styled [textarea element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea). If you are building forms, you may be interested in the TextareaField component instead.
```
```suggestion
Provides a Toucan-styled [input element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input). If you are building forms, you may be interested in the InputField component instead.
```

ynotdraw commented on docs/components/input/index.md:
```diff
@@ -0,0 +1,40 @@
+
+## Value
+
+To set the `value` attribute of the `<textarea>`, provide `@value`.
```
```suggestion
To set the `value` attribute of the `<input>`, provide `@value`.
```

ynotdraw commented on ember-toucan-core/src/components/form/input-field.ts:
```diff
@@ -0,0 +1,39 @@
+    super(owner, args);
+  }
+
+  @action
+  handleInput(e: Event | InputEvent): void {
+    assert('Expected HTMLInputElement', e.target instanceof HTMLInputElement);
+
+    if (this.args.onChange) {
+      this.args.onChange(e.target.value, e);
+    }
+  }
```
Do we still need this here since we can pass the argument directly through to the underlying Input component?

nicolechung replied:
@ynotdraw remove the `handleInput` (or the entire backing class) you mean?

nicolechung replied:
oh wait I see...going to remove the backing class.

simonihmig commented on test-app/tests/integration/components/input-field-test.gts:
```diff
@@ -0,0 +1,148 @@
+      throw new Error('an input was not detected');
+    }
+    
+    assert.strictEqual(input.value, 'Avocado', 'input has the set @value');
```
I would suggest to use `qunit-dom` here, which would also remove need to manually query the element. See https://github.com/CrowdStrike/ember-toucan-core/pull/36/files#diff-b7acb71cb9e9389fb6732d2b7eb21e52658721f8307ab8c972d27de292a9adb3R130

simonihmig commented on test-app/tests/integration/components/input-test.gts:
```diff
@@ -0,0 +1,84 @@
+    await render(<template>
+      {{! we do not require a label, but instead suggest using Field / TextareaField }}
+      {{! template-lint-disable require-input-label }}
+      <InputControl placeholder="Placeholder text" data-input />
```
Interesting, how does template-lint know this is representing an `<input>`? By the component name? The component could render something totally different, where that rule does not apply... 🤔

ynotdraw replied:
Oh interesting! I had to do the same thing with Textarea - may you're right that it is by the component name. I'd be curious!  I'd also be impressed if it dug into the template to find it was in fact an `<input>`!

nicolechung replied:
>Interesting, how does template-lint know this is representing an <input>? By the component name? The component could render something totally different, where that rule does not apply... 🤔

I think it visits all ElementNodes and looks for any instance of say, `Input` or `input`...if it sees `input` it then checks the parent's attributes.
<img width="1227" alt="Screen Shot 2023-02-27 at 1 25 19 PM" src="https://user-images.githubusercontent.com/771170/221651047-b4ec44c5-ba8b-4dd5-819c-e1156c4359f2.png">


https://github.com/ember-template-lint/ember-template-lint/blob/master/lib/rules/require-input-label.js#L88

simonihmig replied:
Yeah, but our tag here is `InputControl`, which is not part of [this SET](https://github.com/ember-template-lint/ember-template-lint/blob/master/lib/rules/require-input-label.js#L29). So still seems weird to me. 🤔 

Anyway, nothing that should block us here, just curious...

nicolechung replied:
I'm not totally understanding what's going on in the lint rule code, but [in the tests](https://github.com/ember-template-lint/ember-template-lint/blob/master/test/unit/rules/require-input-label-test.js#L74) it looks like it searches for any `input` and checks the surrounding (parent?) tags to see if there is an `label`. With `InputControl` it can tell there is an `input` inside but no label.

simonihmig reviewed (commented):
Looks very good overall! Just some minor things to address...

ynotdraw commented on ember-toucan-core/rollup.config.mjs:
```diff
@@ -24,7 +24,11 @@ export default {
     // "app" tree. Things in here should also be in publicEntrypoints above, but
     // not everything in publicEntrypoints necessarily needs to go here.
-    addon.appReexports(['components/*.js', 'components/form/*.js']),
+    addon.appReexports([
+      'components/*.js',
+      'components/form/*.js',
+      'components/form/controls/*.js',
+    ]),
```
If you rebase, you won't need to adjust the rollup config any longer!

(For ref: https://github.com/CrowdStrike/ember-toucan-core/pull/36/files#diff-ef1c3fa189e4cf0ff2a244e5f3ad25086f8d12247f8a37c66e2dc1a8aa8cebb7)

ynotdraw commented on ember-toucan-core/src/components/form/controls/input.hbs:
```diff
@@ -0,0 +1,8 @@
+<input
+  class="bg-overlay-1 focus:outline-none shadow-focusable-outline focus:shadow-focus-outline block rounded-sm p-1
+    {{if @isDisabled 'text-disabled' 'text-titles-and-attributes'}}"
```
I found out yesterday we need a `@hasError` argument so we can conditionally apply the error shadow.  Something  like this should work:

```suggestion
  class="bg-overlay-1 focus:outline-none focus:shadow-focus-outline block rounded-sm p-1
    {{if @isDisabled 'text-disabled' 'text-titles-and-attributes'}}
    {{if @hasError 'shadow-error-outline' 'shadow-focusable-outline'}}"
```

(For reference:  https://github.com/CrowdStrike/ember-toucan-core/pull/52/files#diff-f86c0f8f074d8a714e4e434f71ef3d3ee3db10702a7260cce69f4143ab2b71f7)

nicolechung replied:
Interesting...are we always going to conditionally apply via a `@hasError` argument or should it be `this.hasError` to account for if there are arguments passed to `@error`?

nicolechung replied:
oh nvm `@error` comes from the InputField component

ynotdraw commented on ember-toucan-core/src/components/form/controls/input.ts:
```diff
@@ -0,0 +1,24 @@
+  Element: HTMLInputElement;
+  Args: {
+    isDisabled?: boolean;
+    onChange?: (value: string, e: Event | InputEvent) => void;
```
I added a re-usable type here if you want to give it a go! 

```suggestion
    onChange?: OnChangeCallback<string>;
```

(Reference: https://github.com/CrowdStrike/ember-toucan-core/pull/36/files#diff-90b0ef8397a4c8dabf3349619492464098ba317ce1542a255dcc35ac8640615f)

ynotdraw commented on test-app/tests/integration/components/input-field-test.gts:
```diff
@@ -0,0 +1,148 @@
+    assert.dom(label).hasText('Label', 'Expected to have label text "label"');
+
+    assert.dom(input).exists('Expected to have input tag rendered');
+    assert.dom(input).hasAttribute('type', 'text');
```
We may want to verify the `id` gets set here as well:

```js
assert.dom('[data-input]').hasAttribute('id');
```

nicolechung replied:
okay I think this is here, but you might want to confirm since there is no "Outdated" label

nicolechung replied:
Okay yeah it's here: https://github.com/CrowdStrike/ember-toucan-core/pull/35/files#diff-b1d74a6d42d39cd9e5f8eb62146064c2694715629721f598278442b2a9249855R31

ynotdraw commented on test-app/tests/integration/components/input-field-test.gts:
```diff
@@ -0,0 +1,148 @@
+    const input = '[data-input]';
+    const error = '[data-error]';
+   
+    const inputElement = document.querySelector('[data-input]');
```
Instead of `document.querySelector` you could use `find`

ynotdraw commented on test-app/tests/integration/components/input-field-test.gts:
```diff
@@ -0,0 +1,148 @@
+    assert.dom(error).hasText('There is an error', 'Expected to have error text "error"');
+
+    if (id) {
+      assert.dom(error).hasAttribute('id', id);
```
yup yup.  One thing I did for Textarea to assert this stuff: https://github.com/CrowdStrike/ember-toucan-core/pull/36/files#diff-b7acb71cb9e9389fb6732d2b7eb21e52658721f8307ab8c972d27de292a9adb3R63

```js
     let error = find('[data-error]');

     assert.dom(error).hasText('Error text');
     assert.dom(error).hasAttribute('id');

     let errorId = error?.getAttribute('id') || '';
     assert.ok(errorId, 'Expected errorId to be truthy');

     let describedby =
       find('[data-textarea]')?.getAttribute('aria-describedby') || '';
     assert.ok(
       describedby.includes(errorId),
       'Expected errorId to be included in the aria-describedby'
     );

     assert.dom('[data-textarea]').hasAttribute('aria-invalid', 'true');
```

Then no conditionals and defaulting to `|| ''` and asserting it's truthy

nicolechung replied:
Question: if both the `errorId` and `describedby` end up being `''`, is that assertion correct? Or that case won't happen because assert.ok tests for truthiness? (is that the "errorId" is truthy and `describedby` is truthy??)

nicolechung replied:
I'm still going to use that snippet btw.

ynotdraw commented on test-app/tests/integration/components/input-field-test.gts:
```diff
@@ -0,0 +1,148 @@
+    assert.dom(hint).hasText('Hint text visible here', 'Expected to have hint text "error"');
+
+    if (id) {
+     assert.dom(hint).hasAttribute('id', id);
```
For reference here's how I handled this in Textarea: https://github.com/CrowdStrike/ember-toucan-core/pull/36/files#diff-b7acb71cb9e9389fb6732d2b7eb21e52658721f8307ab8c972d27de292a9adb3R42

ynotdraw commented on test-app/tests/integration/components/input-field-test.gts:
```diff
@@ -0,0 +1,148 @@
+
+    if (id) {
+      assert.dom(error).hasAttribute('id', id);
+    }
```
May be good to also assert `aria-invalid` is set on the `input`

```suggestion
    }
    assert.dom('[data-input]').hasAttribute('aria-invalid', 'true');
```

ynotdraw commented on test-app/tests/integration/components/input-field-test.gts:
```diff
@@ -0,0 +1,148 @@
+
+    assert.verifySteps(['handleChange']);
+  });
+
```
Not sure if overkill, but I added a test for making sure _both_ hint and error IDs are set properly on `aria-describedby`

(Reference: https://github.com/CrowdStrike/ember-toucan-core/pull/36/files#diff-b7acb71cb9e9389fb6732d2b7eb21e52658721f8307ab8c972d27de292a9adb3R81)

nicolechung replied:
I added it here: https://github.com/CrowdStrike/ember-toucan-core/pull/35/files#diff-b1d74a6d42d39cd9e5f8eb62146064c2694715629721f598278442b2a9249855R108

ynotdraw commented on test-app/tests/integration/components/input-test.gts:
```diff
@@ -0,0 +1,84 @@
+    assert.dom('[data-input]').doesNotHaveClass('text-disabled');
+  });
+
+  test('it disables the textarea using `@isDisabled`', async function (assert) {
```
```suggestion
  test('it disables the input using `@isDisabled`', async function (assert) {
```

ynotdraw commented on test-app/tests/integration/components/input-test.gts:
```diff
@@ -0,0 +1,84 @@
+      .doesNotHaveClass('text-titles-and-attributes');
+  });
+
+  test('it spreads attributes to the underlying textarea', async function (assert) {
```
```suggestion
  test('it spreads attributes to the underlying input', async function (assert) {
```

ynotdraw commented on test-app/tests/integration/components/input-test.gts:
```diff
@@ -0,0 +1,84 @@
+
+  test('it sets the value attribute via `@value`', async function (assert) {
+    await render(<template>
+      {{! we do not require a label, but instead suggest using Field / TextareaField }}
```
```suggestion
      {{! we do not require a label, but instead suggest using Field / InputField }}
```

ynotdraw reviewed (commented):
Apologies for more comments! Changed a few things on the Textarea PR before merging!

nicolechung commented on test-app/app/templates/application.hbs:
```diff
@@ -1,3 +1,45 @@
 
-{{outlet}}
\ No newline at end of file
+{{!-- template-lint-disable require-input-label --}}
```
<img width="592" alt="Screen Shot 2023-02-27 at 2 36 30 PM" src="https://user-images.githubusercontent.com/771170/221666174-c63cc846-c0e6-4beb-9344-e3afd1e8741d.png">

Just note: I added this, but did not install the related styles. If we need that maybe it can be a follow up?

simonihmig commented on docs/components/input/index.md:
```diff
@@ -0,0 +1,40 @@
+# Input 
+
+Provides a Toucan-styled [input element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input). If you are building forms, you may be interested in the TextareaField component instead.
```
```suggestion
Provides a Toucan-styled [input element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input). If you are building forms, you may be interested in the InputField component instead.
```

simonihmig commented on ember-toucan-core/unpublished-development-types/index.d.ts:
```diff
@@ -11,5 +12,6 @@ declare module '@glint/environment-ember-loose/registry' {
     // local entries
     'Form::Field': typeof FieldComponent;
     'Form::Controls::Textarea': typeof TextareaControlComponent;
+    'Form::Controls::Input': typeof InputControlComponent;
```
Btw, in our _internal_ types we can also consume our own public registry, so we don't have to add this twice every time. See https://github.com/CrowdStrike/ember-headless-form/blob/main/packages/ember-headless-form/unpublished-development-types/index.d.ts#L32

nicolechung replied:
I think @ynotdraw PR is going to get merged first, will pull that from there: https://github.com/CrowdStrike/ember-toucan-core/pull/50

ynotdraw replied:
Yup! Should be able to rebase and then not have to modify this file! 🎉

simonihmig commented on test-app/tests/integration/components/input-field-test.gts:
```diff
@@ -0,0 +1,165 @@
+
+    const label = '[data-label]'
+    const input = '[data-input]';
+    const inputId = find(input)?.getAttribute('id') || '';
```
When the input has an ID of an empty string (and having `<label for="">`), then I think this would make this test pass. Better to assert also for the existence of a proper ID:

```suggestion
    const inputId = find(input)?.getAttribute('id');
    
    assert.ok(inputId, 'Expected to have ID');
```

nicolechung replied:
<img width="1023" alt="Screen Shot 2023-02-28 at 12 18 34 PM" src="https://user-images.githubusercontent.com/771170/221928523-a0c2fa35-13ed-4ddb-a089-1d46207349c7.png">

The linter was complaining about the line change on 24 but I added the other suggestion - thank you!

ynotdraw commented on docs/components/input/index.md:
```diff
@@ -0,0 +1,40 @@
+
+## Disabled State
+
+Set the `@isDisabled` argument to disable the `<input>`.
```
Nit: May want to add info about `@hasError` :


```suggestion
Set the `@isDisabled` argument to disable the `<input>`.

## Error State

Set the `@hasError` argument to apply an error box shadow to the `<input>`.
```

ynotdraw commented on ember-toucan-core/src/components/form/controls/input.ts:
```diff
@@ -0,0 +1,25 @@
+    hasError?: boolean;
+    isDisabled?: boolean;
+    onChange?: OnChangeCallback<string>;
+    readonly?: boolean;
```
Nit: it looks like we aren't using this anywhere yet

ynotdraw commented on ember-toucan-core/src/components/form/input-field.hbs:
```diff
@@ -0,0 +1,27 @@
+<div
+  class="flex flex-col {{if @isDisabled 'text-disabled cursor-not-allowed'}}"
```
Is the `cursor-not-allowed` class something we want here?

nicolechung replied:
Will remove it for now...I was thinking we wanted to whole input field to be disabled but probably better to adjust this (if even needed!) when design is more finalized.

BTW (not sure if this is clear or not) `cursor-not-allowed` changes the cursor style, it doesn't hide it. https://tailwindcss.com/docs/cursor

Unlike what the naming suggests...

ynotdraw commented on ember-toucan-core/src/components/form/input-field.ts:
```diff
@@ -0,0 +1,29 @@
+    label: string;
+    hint?: string;
+    isDisabled?: boolean;
+    onChange?: (value: string, e: Event | InputEvent) => void;
```
Nit: To keep this in sync with the underlying component API you could do:

```suggestion
    // import ToucanFormInputControlComponentSignature
    onChange?: ToucanFormInputControlComponentSignature['Args']['onChange'];
```

or

```suggestion
    // import OnChangeCallback
    onChange?: OnChangeCallback<string>;
```

ynotdraw commented on ember-toucan-core/src/components/form/input-field.ts:
```diff
@@ -0,0 +1,29 @@
+    hint?: string;
+    isDisabled?: boolean;
+    onChange?: (value: string, e: Event | InputEvent) => void;
+    readonly?: boolean;
```
Same question here about `readonly` - do we want to support this yet?

ynotdraw commented on test-app/tests/integration/components/input-field-test.gts:
```diff
@@ -0,0 +1,166 @@
+    const input = '[data-input]';
+    const error = '[data-error]';
+  
+    assert.dom(label).exists('Expected to have label component rendered');
+    assert.dom(label).hasText('Label', 'Expected to have label text "Label"');
```
Nit: Up to you, but there's a test above validating `label`, so you could probably get away with this test focusing on only testing that the error displays and removing these assertions as they're redundant (same for other tests).

ynotdraw commented on test-app/tests/integration/components/input-field-test.gts:
```diff
@@ -0,0 +1,166 @@
+    assert.dom(label).exists('Expected to have label component rendered');
+    assert.dom(label).hasText('Label', 'Expected to have label text "Label"');
+
+    assert.dom(input).exists('Expected to have input tag rendered');
```
Similar note here, but totally up to you.  The `it renders` test above asserts that the input is there, so keeping the test focused on error / hint states would make the tests smaller and a bit easier to follow potentially.

ynotdraw commented on test-app/tests/integration/components/input-field-test.gts:
```diff
@@ -0,0 +1,166 @@
+
+    assert.verifySteps([]);
+
+    const input: HTMLInputElement | null  = find('[data-input]') as HTMLInputElement;
+
+    if (!input) {
+      throw new Error('an input was not detected');
+    }
+   
+    assert.dom(input).hasValue('Avocado', 'input has the original value');
+
+    await fillIn('[data-input]', 'Banana');
+
+    assert.dom(input).hasValue('Banana', 'input has the set @value');
+
+    assert.verifySteps(['handleChange']);
```
Nit / no need to change: this could maybe be simplified with:

```suggestion
    assert.dom('[data-input]').hasValue('Avocado', 'input has the original value');

    await fillIn('[data-input]', 'Banana');

    assert.dom('[data-input]').hasValue('Banana', 'input has the set @value');

    assert.verifySteps(['handleChange']);
```

If `fillIn` doesn't find an element or if the value isn't set as expected you'd get an assertion error

ynotdraw commented on test-app/tests/integration/components/input-field-test.gts:
```diff
@@ -0,0 +1,166 @@
+    assert.verifySteps(['handleChange']);
+  });
+
+});
```
Nit: May be good as well to add a test for the `rootTestSelector` to make sure we have tests for all of our args:

```suggestion
    test('it applies the provided `@rootTestSelector` to the data-root-field attribute', async function (assert) {
    await render(<template>
      <InputField
        @label="Label"
        @rootTestSelector="selector"
        data-input
      />
    </template>);

    assert.dom('[data-root-field="selector"]').exists();
  });
});
```

ynotdraw reviewed (approved):
🚀 nice work!
